### PR TITLE
Impl. of OpenID Connect token-based API access (a.k.a. DAT access with Keycloak)

### PR DESCRIPTION
**Pull request by pvannierop**
# Background
cBioPortal provides tokens, also referred to as Data Access Tokens or DATs, that allow programmatic access the cBioPortal API. These tokens, as well as user permissions, are managed by the cBioPortal instance.

# Problem
cBioPortal can be configured to use an external Identity Provider (IDP), such as Keycloak, that handles user authentication and user authorization. This setup is not supported by current DAT implementation.

# Solution
We have extended the backend and frontend so that cBioPortal can:
1. Provide an OpenID Connect _offline token_ (=OAuth2 refresh token with scope _offline_access_) to the user. For more information see [OpenID Connect docs](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess).
2. Provide programmatic access to the API with the _offline token_.  

Note: this PR to the backend requires changes to the cbioportal-frontend in [PR #2982](https://github.com/cBioPortal/cbioportal-frontend/pull/2982).

# Details
- The implementation from the viewpoint of the user is identical to the existing DAT implementation; no changes to the user experience.
- Although the _offline token_ has a near infinite lifespan, the user permissions are fetched from the OpenID Connect client upon each request to the cBioPortal API. Changes in user permissions at the side of the IDP are instantaneous and do not require invalidation of the token. _Offline tokens_ can be revoked at the level of the IDP at any time; removal of the token is effective immediately.
- Responsibility for triggering the download of the token has been moved to the cBioPortal backend (is handled by the frontend at present).
- Integration test have been added that validate expected responses of the API endpoints when accessed with valid, invalid or no tokens (SAML/OAuth2 only).
- Documentation has been updated to include instructions on the configuration of the OpenID Connect client in Keycloak.

### CODE REVIEW

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md [hunk 1]
```diff
@@ -252,6 +253,85 @@ If all goes well, the following should happen:
 
 If this does not happen, see the Troubleshooting Tips  below.
 
+### Add client for OAuth2 token-based data access
+With cBioPortal instances that require user authentication the API can be queried when including a data access tokens in the request header (see [Authenticating Users via Tokens](Authenticating-Users-via-Tokens.md)). When using the `oauth2` implementation, KeyCloak can be configured as OAuth2 authentication provider that distributes data access tokens to users and validates these tokens when used while querying the API. This feature is enabled by creating a `cbioportal_api` OpenID Connect client that has access to the user roles defined in the `cbioportal` SAML client.
```
Comment (n1zea144) [inline docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: Suggest rewording  in first sentence - s/tokens/token/
Suggest rewording in second sentence - "Keycloak can be configured as an OAuth2..."

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md [hunk 2]
```diff
@@ -252,6 +253,85 @@ If all goes well, the following should happen:
...
+
+1. Create a client with name `cbioportal_api`. Set _Client Protocol_ to `openid-connect`.
+
+![](images/previews/oauth2_client_1.png)
+
+2. On the configuration page of `cbioportal_api` client apply the following settings:
```
Comment (n1zea144) [inline docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: Nice summary

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md [hunk 3]
```diff
@@ -252,6 +253,85 @@ If all goes well, the following should happen:
...
+| Standard Flow Enabled | ON      |     (default value) |
+| Implicit Flow Enabled | OFF      |     (default value) |
+| Direct Access Grants Enabled | OFF      |    |
+| Service Accounts Enabled | OFF      |  (default value)   |
+| Authorization Enabled | OFF      |  (default value)  |
+| Valid Redirect URIs | _url_/api/data-access-token/oauth2  |  _url_ refers to base url of cBioPortal instance |
+
+![](images/previews/oauth2_client_3.png)
+
+#### Credentials tab
```
Comment (n1zea144) [inline docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: This may be version specific - running keycloak 9, I don't see this tab

image

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md [hunk 4]
```diff
@@ -252,6 +253,85 @@ If all goes well, the following should happen:
...
+
+![](images/previews/oauth2_client_5.png)
+
+3. Add these parameters to `portal.properties` of the cBioPortal backend .
+
+| parameter        | value  | comment  |
+| ------------- |:-------------:| -----:|
+| dat.oauth2.clientId       | cbioportal_api |  |
+| dat.oauth2.clientSecret    | ?      | see _Secret_ field in the _Credentials_ tab  |
+| dat.oauth2.accessTokenUri     | _url_/auth/realms/cbio/protocol/openid-connect/token      |   _url_ refers to base url of the KeyCloak server |
```
Comment (n1zea144) [inline docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: In this case isn't cbio is the name of the realm you created? I think the docs use cbioportal, but maybe you should use _realm-name_ or something along those lines.  Relevant for some values below...

File: docs/Authenticating-Users-via-Tokens.md [hunk 5]
```diff
@@ -27,20 +28,27 @@ mysql> describe data_access_tokens;
...
+
+A step-by-step guide to configure KeyCloak to provide OAuth2 client functionality can be found in [Authenticating and Authorizing Users via keycloak](Authenticating-and-Authorizing-Users-via-keycloak.md#Add-client-for-OAuth2-token-based-data-access).
+
 ### Modifying Configuration
-The following properties must be present in portal.properties in order to allow direct access to the cBioPortal web service when login is required:
+The token access implementation can be specified with the `dat.method` Java parameter at startup of cBioPortal. Allowed values are `jwt`, `uuid`, `oauth2` and `none`. When no parameter is specified, `none` will be used as default value and access using tokens is not possible.
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply docs/Authenticating-Users-via-Tokens.md]: Indeed I implemented switching between dat.method implementations using a spring active profile. I have removed this method using a conditional bean creation using `@Conditional` annotations. A special spring boot param is no longer required. Start of cbioportal is same as was before.

File: docs/Authenticating-Users-via-Tokens.md [hunk 6]
```diff
@@ -60,6 +68,36 @@ The following properties must be present in portal.properties in order to allow
 - **Permissible Values**: true or false
 - **Default Value**: true
 
+**Property**: dat.oauth2.clientId (required only when dat.method = oauth2)
+- **Description**: Identifier of the OAuth2 client of the authentication provider. This value is specified by the authentication provider.
+- **Permissible Values**: string
+- **Default Value**:
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply docs/Authenticating-Users-via-Tokens.md]: All properties were updated as per your suggestion.

File: docs/Authenticating-Users-via-Tokens.md [hunk 7]
```diff
@@ -69,15 +107,15 @@ This will bring you to a section of the Web API page where you can download your
 
 ![](images/previews/download-dat-button.png)
 
-The downloaded data access token file will contain the token, a creation date, and an expiration date:
+The downloaded data access token file will contain the token, and, depending on the implementation, a creation and expiration date:
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply docs/Authenticating-Users-via-Tokens.md]: Yes, line was made more specific.

File: docs/Authenticating-Users-via-Tokens.md [hunk 8]
```diff
@@ -27,20 +28,27 @@ mysql> describe data_access_tokens;
...
 
+A third implementation makes use an external OAuth2-authentication provider and can integrate with single-sign on (SSO) solutions. The OAuth2 provider provides users with an __offline token__ that does not allow for direct access to protected cBioPortal resources. Instead, the offline token is passed to the OAuth2 authorization server that returns an __access token__, a short-lived permission to access cBioPortal resources. The access token contains up-to-date user permissions (the studies that the user is permitted to view). This implemetation ensures that any changes of user permissions registered at the authentication provider are effective immediately and do not relie on the expiration date of the token of the user. At any moment OAuth2 offline tokens can be revoked by the Authentication provider on a per-user basis.
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply docs/Authenticating-Users-via-Tokens.md]: Typo's corrected

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md (hunk 1 above)
Comment (pvannierop) [reply docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: Updated

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md (hunk 4 above)
Comment (pvannierop) [reply docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: Well spotted. I have updated `/cbio/` to `/cbioportal/`.

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md (hunk 3 above)
Comment (pvannierop) [reply docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: I am running 8 and it is there...

File: portal/src/integration-tests/saml-oauth2-setup/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+- `OfflineTokenDownloadIntegrationTests` tests downloading an offline token to gain access to cioportal API.
+- `Oauth2ResourceServerIntegrationTests` tests API access by different type of users (anonymous, non authorised and authorised).
+
+## Start web container with cbioportal and saml IDP
+
+You can start tomcat to play with the setup with the following command (replace with your values):
+
+```$bash
+ CBIO_TEST_DB_USR=<user> CBIO_TEST_DB_PSW=<psw> CBIO_TEST_DB_HOST=127.0.0.1:3306 CBIO_TEST_DB_NAME=cgds_test CBIO_TEST_DB_CONNECTION_STRING=jdbc:mysql://127.0.0.1:3306/cgds_test?sessionVariables=default_storage_engine=InnoDB CBIO_WAR_LOCATION=../../../target/cbioportal.war mvn org.codehaus.cargo:cargo-maven2-plugin:run --non-recursive
```
[the remark this answers is not in the document]
Comment (inodb) [reply portal/src/integration-tests/saml-oauth2-setup/README.md]: same: not familiar with cargo either, but would indeed be nice to be closest as possible to production setup (i.e. webapp-runner) if possible

File: docs/Authenticating-Users-via-Tokens.md [hunk 10]
```diff
@@ -27,20 +28,27 @@ mysql> describe data_access_tokens;
...
 
+A third implementation makes use an external OAuth2-authentication provider and can integrate with single-sign on (SSO) solutions. The OAuth2 provider provides users with an __offline token__ that does not allow for direct access to protected cBioPortal resources. Instead, the offline token is passed to the OAuth2 authorization server that returns an __access token__, a short-lived permission to access cBioPortal resources. The access token contains up-to-date user permissions (the studies that the user is permitted to view). This implemetation ensures that any changes of user permissions registered at the authentication provider are effective immediately and do not relie on the expiration date of the token of the user. At any moment OAuth2 offline tokens can be revoked by the Authentication provider on a per-user basis. 
+
+A step-by-step guide to configure KeyCloak to provide OAuth2 client functionality can be found in [Authenticating and Authorizing Users via keycloak](Authenticating-and-Authorizing-Users-via-keycloak.md#Add-client-for-OAuth2-token-based-data-access).
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply docs/Authenticating-Users-via-Tokens.md]: Thnx :)

File: portal/src/integration-tests/saml-oauth2-setup/README.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# saml-oauth2-setup
+
+Tests oauth2 security workflow of the API with SAML authentication method for the UI.
+
+- `OfflineTokenDownloadIntegrationTests` tests downloading an offline token to gain access to cioportal API.
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply portal/src/integration-tests/saml-oauth2-setup/README.md]: Typo corrected.

File: portal/src/integration-tests/saml-oauth2-setup/README.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+- `Oauth2ResourceServerIntegrationTests` tests API access by different type of users (anonymous, non authorised and authorised).
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply portal/src/integration-tests/saml-oauth2-setup/README.md]: Corrected to 'z' :)

File: portal/src/integration-tests/saml-oauth2-setup/pom.xml [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply portal/src/integration-tests/saml-oauth2-setup/pom.xml]: Added.

File: portal/src/integration-tests/saml-oauth2-setup/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (pvannierop) [reply portal/src/integration-tests/saml-oauth2-setup/README.md]: The cargo plugin is configured to run Tomcat8 in the portal/src/integration-tests/saml-oauth2-setup/pom.xml:

```
<containerId>tomcat8x</containerId>
```
This feature was developed by Ruslan in our company and he is no longer available for changes. I cannot find any easy change to webapp runnen. I am afraid I have to pass on this request.

File: portal/src/integration-tests/saml-oauth2-setup/pom.xml (hunk 13 above)
[the remark this answers is not in the document]
Comment (pvannierop) [reply portal/src/integration-tests/saml-oauth2-setup/pom.xml]: Done.

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md (hunk 2 above)
Comment (pvannierop) [reply docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: Tnnx.

File: portal/src/integration-tests/saml-oauth2-setup/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (n1zea144) [reply portal/src/integration-tests/saml-oauth2-setup/README.md]: If understand the web app runner setup, it looks like it is configured to run 8.5.37 version of tomcat, so I think we are good here (I must have missed the tomcat8x containerId)

File: portal/src/integration-tests/saml-oauth2-setup/pom.xml [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+                            l
+                            <saml.logout.local>false</saml.logout.local>
+                            <!--FIXME Our test saml idp does not sing assertions for some reason-->
+                            <saml.sp.metadata.wantAssertionSigned>false</saml.sp.metadata.wantAssertionSigned>
+                            <saml.logout.url>/</saml.logout.url>
+                            <!-- OAuth2 settings -->
+                            <dat.oauth2.clientId>client_id</dat.oauth2.clientId>
+                            <dat.oauth2.clientSecret>client_secret</dat.oauth2.clientSecret>
+                            <dat.oauth2.issuer>token_issuer</dat.oauth2.issuer>
+                            <dat.oauth2.accessTokenUri>http://localhost:8443/auth/realms/cbio/token
```
[the remark this answers is not in the document]
Comment (n1zea144) [reply portal/src/integration-tests/saml-oauth2-setup/pom.xml]: I now understand that an actually keycloak server is not run during tests, but "real" URLs are still intercepted.  Can we change the realm here from cbio to cbioportal just for consistency?

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md [hunk 15]
```diff
@@ -252,11 +253,98 @@ If all goes well, the following should happen:
...
+#### Settings tab
+
+| parameter        | value  | comment  |
+| ------------- |:-------------:| -----:|
+| Access Type      | confidential |  |
+| Standard Flow Enabled      | ON      |   (default value) |
+| Enabled      | ON      |   (default value) |
+| Consent Required      | OFF        |   (default value) |
+| Client Protocol      | openid-connect   |   (default value) |
+| Access Type | confidential      |     (default value) |
```
Comment (n1zea144) [inline docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: Is this duplication of line 272?

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md [hunk 16]
```diff
@@ -252,11 +253,98 @@ If all goes well, the following should happen:
...
+| Direct Access Grants Enabled | OFF      |    |
+| Service Accounts Enabled | OFF      |  (default value)   |
+| Authorization Enabled | OFF      |  (default value)  |
+| Valid Redirect URIs | _url_/api/data-access-token/oauth2  |  _url_ refers to base url of cBioPortal instance |
+
+![](images/previews/oauth2_client_3.png)
+
+#### Credentials tab
+
+Select `Client Id and Secret`. Take notice of the value of _Secret_ the secret field. This secret should be added to `portal.properties` file of the cBioPortal backend.
```
Comment (n1zea144) [inline docs/Authenticating-and-Authorizing-Users-via-keycloak.md]: It would be helpful to add a link to line 74 of [Authenticating-Users-via-Tokens.md](https://github.com/cBioPortal/cbioportal/blob/fd3f6e5111ffed07d6eba8975910bdc52743d9e2/docs/Authenticating-Users-via-Tokens.md)

File: web/src/main/java/org/cbioportal/web/config/DataAccessTokenControllerConfig.java [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+import org.cbioportal.web.DataAccessTokenController;
+import org.cbioportal.web.OAuth2DataAccessTokenController;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.context.annotation.PropertySources;
+
+@Configuration
+@PropertySources({
```
Comment (n1zea144) [inline web/src/main/java/org/cbioportal/web/config/DataAccessTokenControllerConfig.java]: I think this needs some tweaking.  Since I deploy war with portal.properties rolled in, to avoid a FileNotFoundException, I had to do the following:

> @PropertySource(value="file:///${portal_home}/portal.properties", ignoreResourceNotFound=true)

I imagine you want to look in war first and if not found, then environment?  Maybe you need to place ignoreResourceNotFound on both @PropertySource and then do something smart in the condition class?

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/token/oauth2/OAuth2DataAccessTokenServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    @Value("${dat.oauth2.clientSecret:}")
+    private String clientSecret;
+
+    @Value("${dat.oauth2.accessTokenUri:}")
+    private String accessTokenUri;
+
+    @Value("${dat.oauth2.userAuthorizationUri:}")
+    private String userAuthorizationUri;
+
+    @Value("${dat.oauth2.redirectUri:}")
```
Comment (n1zea144) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/token/oauth2/OAuth2DataAccessTokenServiceImpl.java]: Setting this is missing from the documents.

**Comment by pvannierop** [comment]
@n1zea144 Omission has been fixzed in new commit.

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/token/oauth2/OAuth2DataAccessTokenServiceImpl.java (hunk 18 above)
Comment (pvannierop) [reply security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/token/oauth2/OAuth2DataAccessTokenServiceImpl.java]: Fixed.

File: web/src/main/java/org/cbioportal/web/config/DataAccessTokenControllerConfig.java (hunk 17 above)
Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/config/DataAccessTokenControllerConfig.java]: I have updated to this:
```
@Configuration
@PropertySources({
    @PropertySource("file:///${PORTAL_HOME}/portal.properties", ignoreResourceNotFound=true),
    @PropertySource("classpath:portal.properties", ignoreResourceNotFound=true)
})
```

I do not think the Condition classes need to be updated since these check for presence of the `dat.method` prop. If not present, the beans are not created without making a fuss.

File: docs/Authenticating-Users-via-Tokens.md [hunk 19]
```diff
@@ -60,7 +66,43 @@ The following properties must be present in portal.properties in order to allow
...
+**Property**: dat.oauth2.jwkUrl (required only when dat.method = oauth2)
+- **Description**: Url for retrieval of the public key used by the OAuth2 authorization provider to sign tokens. According to OAuth2 specification this endpoint must end with `/certs`.
+- **Permissible Values**: url
+- **Default Value**: This value is specified by the authentication provider.
+
+**Property**: dat.oauth2.jwtRolesPath (required only when dat.method = oauth2)
+- **Description**: `::`-separated path to array with user roles in JWT token returned by Keycloak.
+- **Permissible Values**: `::`-separated strings
+- **Default Value**: resource_access::cbioportal::roles
+
```
Comment (n1zea144) [inline docs/Authenticating-Users-via-Tokens.md]: I think you are missing dat.oauth2.redirectUri

File: web/src/main/java/org/cbioportal/web/DataAccessTokenController.java [hunk 20]
```diff
@@ -1,98 +1,97 @@
...
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.client.HttpClientErrorException;
 
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiParam;
+
 @InternalApi
-@RestController
+@RequestMapping // replaces @RestController; controller is created conditionally in DataAccessTokenControllerConfig
```
Comment (n1zea144) [inline web/src/main/java/org/cbioportal/web/DataAccessTokenController.java]: It looks like the removal of @RestController causes the following DataAccessTokenControllerTest.revokeAllTokensForUserTest() failure:

`
Running org.cbioportal.web.DataAccessTokenControllerTest
Tests run: 7, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 17.534 sec <<< FAILURE!
revokeAllTokensForUserTest(org.cbioportal.web.DataAccessTokenControllerTest)  Time elapsed: 0.772 sec  <<< ERROR!
javax.servlet.ServletException: Circular view path [data-access-tokens]: would dispatch back to the current handler URL [/data-access-tokens] again. Check your ViewResolver setup
! (Hint: This may be the result of an unspecified view, due to default view name generation.)      
`

After some digging, I think its because @RestController included @ResponseBody which is required to properly unmarshal into JSON.  Adding @ResponseBody after @RequestMapping should solve this issue.

Comment (pvannierop) [reply web/src/main/java/org/cbioportal/web/DataAccessTokenController.java:52]: PR was updated with this solution.

**Comment by n1zea144** [comment]
Just a quick update.  This PR has been reviewed and approved for new functionality.  Whats remaining is a regression test to ensure existing token functionality continues to work.  This task will be put on pipeline scrum starting on May 27.

**Comment by ao508** [comment]
@pvannierop I have been attempting to deploy a portal for regression testing based off of this branch and I'm encountering this error:

```
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'otherController': Unsatisfied dependency expressed through field 'tokenService'; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type 'org.cbioportal.service.DataAccessTokenService' available: expected single matching bean but found 4: uuid,none,jwt,oauth2
```

I remember encountering something similar when we were initially developing this feature and I resolved it through setting up a bean factory that would set up the `tokenService` bean based on the type of `dat.method` set. I see the factory method for resolving which of these to use is gone though and replaced with a config component that conditionally?

Were you able to regression test this code with the JWT/UUID methods? I'm wondering if this might be isolated to my current environment or if you did not have any issue deploying portals with these other DAT methods.

File: docs/Authenticating-Users-via-Tokens.md (hunk 19 above)
Comment (pvannierop) [reply docs/Authenticating-Users-via-Tokens.md:103]: Fixed

File: security/security-spring/src/test/java/org/cbioportal/security/spring/authentication/token/oauth2/OAuth2DataAccessTokenServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+                .append("\"iss\":\"").append("invalid_issuer").append("\"")
+            .append("}")
+            .toString();
+
+        String token = createJwt(json);
+
+        service.isValid(token);
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (pvannierop) [reply security/security-spring/src/test/java/org/cbioportal/security/spring/authentication/token/oauth2/OAuth2DataAccessTokenServiceImplTest.java:159]: Done

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md (hunk 15 above)
Comment (pvannierop) [reply docs/Authenticating-and-Authorizing-Users-via-keycloak.md:277]: Yes, removed

File: docs/Authenticating-and-Authorizing-Users-via-keycloak.md (hunk 16 above)
Comment (pvannierop) [reply docs/Authenticating-and-Authorizing-Users-via-keycloak.md:289]: Link added.

**Comment by ao508** [comment]
@pvannierop When I attempt to download the file that stores your data access token info I get redirected to a different page instead of triggering a file download. On the other hand, I'm no longer getting the exception I was getting before where the DAT service bean couldnt be resolved.

**Comment by n1zea144** [comment]
@pvannierop It sounds like this backend PR can be merged pending the identification of the issue with downloading a token from the frontend.  I understand from @ao508 you are working on this.  Do you want us to wait to approve this PR until your frontend testing is complete?

**Comment by pvannierop** [comment]
@n1zea144 Please do NOT merge. This is an API changing PR that should go in at the same time as the frontend commit.

**Comment by n1zea144** [comment]
Per @ao508 regression tests pass, so ok to approve this PR.
